llvm/lib/Transforms/Utils/CodeLayout.cpp

```diff
@@ -753,15 +774,23 @@ class ExtTSPImpl {
             BestGain = CurGain;
             BestChainPred = ChainPred;
             BestChainSucc = ChainSucc;
+            // Stop early when the merge is as good as the previous one.
```

Comment:
If PrevScore is changed to a std::optional, add an `assert(!PrevScore || BestGain.score() > *PrevScore + EPS)`